micro-rdk/src/esp32/nvs_storage.rs

```diff
@@ -207,13 +214,44 @@ impl RobotConfigurationStorage for NVSStorage {
     }
 
     fn store_robot_configuration(&self, cfg: &RobotConfig) -> Result<(), Self::Error> {
-        self.set_blob(NVS_ROBOT_CONFIG_KEY, cfg.encode_to_vec().into())?;
+        // Normally, we would just rely on blob comparison in
+        // `NVSStorage::set_blob` to dedup the write, but if snappy
+        // compression of the robot config isn't entirely
+        // deterministic, then that protection would be defeated. So,
+        // check the preimages here instead, and if they are
+        // equivalent, skip writing.
+        if self.has_robot_configuration() {
+            if let Ok(cached_cfg) = self.get_robot_configuration() {
+                if cached_cfg == *cfg {
+                    return Ok(());
+                }
+            }
+        }
+
+        let encoded_cfg = cfg.encode_to_vec();
+        let mut compressor = FrameEncoder::new(Vec::new());
+        compressor
+            .write_all(&encoded_cfg[..])
+            .map_err(NVSStorageError::NVSRobotConfigCompressionError)?;
+        let compressed = compressor
+            .into_inner()
+            .map_err(|e| NVSStorageError::NVSRobotConfigCompressionError(e.into_error()))?;
+        log::info!(
+            "robot configuration compressed for NVS storage: {} bytes before, {} bytes after",
+            encoded_cfg.len(),
+            compressed.len()
+        );
+        self.set_blob(NVS_ROBOT_CONFIG_KEY, compressed.into())?;
         Ok(())
     }
 
     fn get_robot_configuration(&self) -> Result<RobotConfig, Self::Error> {
-        let robot_config = self.get_blob(NVS_ROBOT_CONFIG_KEY)?;
-        RobotConfig::decode(&robot_config[..]).map_err(NVSStorageError::NVSValueDecodeError)
+        let mut decompressed_robot_config = vec![];
```

Comment:
One thing we could do to harden this here would be that if we fail to decompress the buffer with snappy, that we then try to just feed the raw blob `RobotConfig::decode`. That'd protect against a situation where you had a non-compressed robot config in NVS, got OTA updated to a version that expected a compressed robot config, and then for some reason when you came back online were unable to reach app and the expectation would be to use the cached config.